pkg/sql/sessioninit/cache.go

```diff
@@ -159,16 +164,23 @@ func (a *Cache) GetAuthInfo(
 				return nil
 			}
 
-			// Lookup the data outside the lock.
-			aInfo, err = readFromSystemTables(
-				ctx,
-				txn,
-				ie,
-				username,
+			// Lookup the data outside the lock, with at most one request in-flight
+			// for each user.
+			aInfoInterface, _, err := a.populateCacheGroup.Do(
+				username.Normalized(),
+				func() (interface{}, error) {
+					return readFromSystemTables(
+						ctx,
+						txn,
+						ie,
+						username,
+					)
+				},
 			)
 			if err != nil {
 				return err
 			}
+			aInfo = aInfoInterface.(AuthInfo)
```

Comment:
I think this needs to use `DoChannel` so you can get canceled. Also, fork a new context with the log tags of this context but based on background and is canceled on stop so that if the query gets canceled, all the other queries don't get a super awkward context canceled error.

Reply:
done